src/Build/Graph/ProjectGraph.cs

```diff
@@ -54,5 +59,139 @@ public ProjectGraph(string entryProjectFile)
+                ProjectGraphNode node = nodeWithTargets.Node;
+                IList<string> targets = nodeWithTargets.Targets;
+
+                targetLists[node].AddRange(targets);
+
+                // No need to continue if this node has no project references.
+                if (node.ProjectReferences.Count == 0)
+                {
+                    continue;
+                }
+
+                // Based on the entry points of this project, determine which targets to propagate down to project references.
+                var targetsToAppendToChildren = new List<string>();
+                ICollection<ProjectItem> projectReferenceTargets = node.Project.GetItems(ProjectReferenceTargetsItemType);
+                foreach (var target in targets)
+                {
+                    // TODO: Normalize targets? Default targets needs expansion. Flip the loop maybe
+
+                    foreach (var projectReferenceTarget in projectReferenceTargets)
+                    {
+                        if (projectReferenceTarget.EvaluatedInclude.Equals(target, StringComparison.OrdinalIgnoreCase))
+                        {
+                            string targetsMetadataValue = projectReferenceTarget.GetMetadataValue(ProjectReferenceTargetsMetadataName);
+                            targetsToAppendToChildren.AddRange(ExpressionShredder.SplitSemiColonSeparatedList(targetsMetadataValue));
+                        }
+                    }
+                }
+
+                // Queue the project references for visitation, if the edge hasn't already been traversed.
+                foreach (var projectReference in node.ProjectReferences)
+                {
+                    var projectReferenceEdge = new ProjectGraphNodeWithTargets(projectReference, targetsToAppendToChildren);
+                    if (visitedEdges.Add(projectReferenceEdge))
+                    {
+                        edgesToVisit.Enqueue(projectReferenceEdge);
+                    }
+                }
+            }
+
+            // Dedupe target lists.
+            foreach (KeyValuePair<ProjectGraphNode, List<string>> pair in targetLists)
+            {
+                List<string> targetList = pair.Value;
+                HashSet<string> seenTargets = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                int i = 0;
+                while (i < targetList.Count)
+                {
+                    if (seenTargets.Add(targetList[i]))
+                    {
+                        i++;
+                    }
+                    else
+                    {
+                        targetList.RemoveAt(i);
+                    }
+                }
+            }
+
+            // Massage the final dictionary into string arrays instead of lists.
+            return targetLists.ToDictionary(pair => pair.Key, pair => pair.Value.ToArray());
+        }
+
+        private struct ProjectGraphNodeWithTargets : IEquatable<ProjectGraphNodeWithTargets>
```

Comment:
This actually represents an edge, so should it be called something like `IncomingTargetsEdge` instead? It would make it fit in better with the `edge` named local variables.